modules/microvm/virtualization/microvm/appvm.nix

```diff
@@ -120,6 +121,64 @@
     autostart = true;
     config = appvmConfiguration // {imports = appvmConfiguration.imports ++ cfg.extraModules ++ vm.extraModules ++ [{environment.systemPackages = vm.packages;}];};
   };
+
+  formatHexOctet = n:
+    assert 0 <= n && n < 256; let
+      n' = lib.trivial.toHexString n;
+      # Pad with zero.
+      # Length is always 1 or 2, because of assertion.
+      prefix =
+        if stringLength n' == 1
+        then "0"
+        else "";
+    in
+      prefix + n';
+
+  allocateMacsUnsafe = vms: let
+    takenMacAddresses = filter (macAddress: macAddress != null) (map (vm: vm.macAddress) vms);
+
+    appvmMacAddressSpace = map (i: "02:00:00:03:${formatHexOctet i}:01") (lib.lists.range 0 255);
+
+    availableMacAddresses = lib.lists.subtractLists takenMacAddresses appvmMacAddressSpace;
+  in
+    (foldl' ({
+        availableMacAddresses,
+        vms,
+      }: vm: let
+        allocated =
+          if vm.macAddress == null
```

Comment:
This works overlaps with https://github.com/tiiuae/ghaf/pull/668/files#r1651177047